### ISSUE TITLE
Optimize performance / energy use of GitHub actions

### ISSUE THREAD

**Issue by dabrahams** [hylo-lang/hylo#951]
- Documentation extraction depends on the same build steps that the regular build/test does, use [dependent jobs](https://www.edwardthomson.com/blog/github_actions_17_dependent_jobs.html) to do the build steps once for both cases
- LLVM setup is an early build step that is depended on by build steps for different configurations on a given platform; the same kind of refactor could be used
- Evaluate the relative time it takes to do things on the different platforms, including the devcontainers, and consider choosing the platform for doc extraction differently.
- Consider whether there's anything we can do to use cacheing (actions/cache@v3) more effectively, e.g. cacheing the results of LLVM builds.

See also #950, which is related.

**Comment by WalterSmuts** [comment]
> * Documentation extraction depends on the same build steps that the regular build/test does, use [dependent jobs](https://www.edwardthomson.com/blog/github_actions_17_dependent_jobs.html) to do the build steps once for both cases
> * LLVM setup is an early build step that is depended on by build steps for different configurations on a given platform; the same kind of refactor could be used

You can't simply use dependent jobs to ingest build artifacts from other jobs. See [this stackoverflow question](https://stackoverflow.com/questions/57498605/github-actions-share-workspace-artifacts-between-jobs) for proof. The suggestion is to add the dependency *and* transfer the artefacts with `actions/upload-artifact` and `actions/download-artifact`. Since the entire setup is downloading and installing, the former should already be cached and the latter cannot be cached, there is not really anything to be gained here.

For the LLVM artefacts we can cache the downloading of the binaries but not the installation on the specific platform. See [documentation[(https://github.com/KyleMayes/install-llvm-action#example-usage-with-caching). This would be worth doing.

**Comment by dabrahams** [comment]
Any alternative is fine with me.  The use of dependent jobs wasn't really supposed to be a part of the goal, even if it was phrased that way.

**Comment by dabrahams** [comment]
Also it would be good to evaluate our existing cache steps to see if they are providing any value.

**Comment by dabrahams** [comment]
@WalterSmuts IIRC you found that caches were a net loss, and [this](https://github.com/KyleMayes/install-llvm-action/commit/878985d084d4991346a5f6f26eae447ddc16457a) seems to confirm that.

**Comment by WalterSmuts** [comment]
Yup, it was triggered by https://github.com/KyleMayes/install-llvm-action/issues/56
